### PR TITLE
create nanos instance

### PR DESCRIPTION
**Pull request by tijoytom**

### CODE REVIEW

**Comment by tijoytom** [comment]
`ops  image create -c config.json data/server`
--Wait for the image creation, it's async

`./ops instance new -n <ProjectID> nanos-server-image`
where  nanos-server-image is  the image name created with ops  image create.

Couple of notes.
1. Now it works only on linux, i had to use GNU tar, for some reason  golang  GNU tar is different and compute engine http 500 
2. Image creation and instance creation are fire and forgot, compute engine give you a selflink for the operation to check for status. I know aws has an option to wait for operation to compete, but gcp do not have it. We need to handle this
3. Build a generic framework for send and receiving API messages.

File: lepton/config.go [hunk 1]
```diff
@@ -22,10 +22,10 @@ type Config struct {
...
 
 type ProviderConfig struct {
-	Platform    string `cloud:"platfom"`
-	ProjectID   string `cloud:"projectid"`
-	BucketName  string `cloud:"bucketname"`
-	ArchiveName string `cloud:"arcname"`
+	Platform   string `cloud:"platfom"`
+	ProjectID  string `cloud:"projectid"`
+	BucketName string `cloud:"bucketname"`
+	ImageName  string `cloud:"imageame"`
```
Comment (eyberg) [inline lepton/config.go]: imagename

File: lepton/gcp.go [hunk 2]
```diff
@@ -32,19 +34,20 @@ func (p *GCloud) BuildImage(ctx *Context) (string, error) {
...
 	if err != nil {
 		return "", err
 	}
 
-	archPath := filepath.Join(filepath.Dir(imagePath), c.CloudConfig.ArchiveName)
+	archPath := filepath.Join(filepath.Dir(imagePath), c.CloudConfig.ImageName+".tar.gz")
 	files := []string{symlink}
 
-	err = createArchive(archPath, files)
+	// TODO : createArchiveExternal calls tar and won't work
```
Comment (eyberg) [inline lepton/gcp.go]: I'd WIP this PR until the go tar issue is resolved

File: lepton/gcp.go [hunk 3]
```diff
@@ -32,19 +34,20 @@ func (p *GCloud) BuildImage(ctx *Context) (string, error) {
...
 	}
 
-	err = os.Symlink(imagePath, symlink)
+	err = os.Link(imagePath, symlink)
 	if err != nil {
 		return "", err
 	}
 
-	archPath := filepath.Join(filepath.Dir(imagePath), c.CloudConfig.ArchiveName)
+	archPath := filepath.Join(filepath.Dir(imagePath), c.CloudConfig.ImageName+".tar.gz")
```
Comment (eyberg) [inline lepton/gcp.go]: this is repeated in a few spots - let's declare it somewhere

File: lepton/gcp.go [hunk 4]
```diff
@@ -107,11 +110,18 @@ func (p *GCloud) CreateInstance(ctx *Context) error {
...
+		filepath.Base(c.CloudConfig.ImageName),
+		strconv.FormatInt(time.Now().Unix(), 10),
+	)
 
 	imageName := fmt.Sprintf("projects/%v/global/images/%v",
 		c.CloudConfig.ProjectID,
-		strings.Split(c.CloudConfig.ArchiveName, ".")[0])
+		c.CloudConfig.ImageName)
+
+	fmt.Println(machineType)
```
Comment (eyberg) [inline lepton/gcp.go]: debug output?

File: lepton/gcp.go [hunk 5]
```diff
@@ -139,6 +149,12 @@ func (p *GCloud) CreateInstance(ctx *Context) error {
 	return nil
 }
 
+func createArchiveExternal(archive string, files []string) error {
+	cmd := exec.Command("tar", "-czvf", archive, filepath.Base(files[0]))
```
Comment (eyberg) [inline lepton/gcp.go]: as mentioned above - need to figure out what's going on w/native go tar and should prob. use that instead

File: ops.go [hunk 6]
```diff
@@ -258,11 +258,11 @@ func setDefaultImageName(cmd *cobra.Command, c *api.Config) {
 		imageName = path.Join(images, filepath.Base(imageName))
 	}
 	c.RunConfig.Imagename = imageName
-	c.CloudConfig.ArchiveName = fmt.Sprintf("nanos-%v-image.tar.gz", filepath.Base(c.Program))
+	c.CloudConfig.ImageName = fmt.Sprintf("nanos-%v-image", filepath.Base(c.Program))
```
Comment (eyberg) [inline ops.go]: this is used elsewhere - should prob. make it declared

**Comment by eyberg** [comment]
also - besides the tar thing i'd also WIP this until we get a spinner on the 'Instance creation started.' to whatever status says it booted or it failed && also block until we get that new nanos release out the door

**Comment by ssudake21** [comment]
@eyberg @tijoytom 
Fixed couple of things
1. Add default Empheral Public IP for instance
2. Open HTTP, HTTPS port for Instance
3.  Enable Serial port connection for Instance
4. Monitor Instance/Image creation status until success/failure

**Pull request merged**